app/models/finance/postpaid_billing_strategy.rb

```diff
@@ -35,13 +35,20 @@ def description
   end
 
   def bill_plan_change(contract, period)
-    if contract.paid_until.to_date < period.begin.to_date
-      add_plan_cost(:bill, contract, contract.plan, period)
-    else
-      # FIXME: replace end_of_month by paid until!
-      add_plan_cost(:refund, contract, contract.old_plan, period)
-      add_plan_cost(:bill, contract, contract.plan, period)
+    plan = contract.plan
+    old_plan = contract.old_plan
+    period_begin = period.begin.utc
+    period_end = period.end.utc
+    paid_until = contract.paid_until.utc
+
+    if paid_until < period_begin # This is only for fixed cost, so it means billing didn't run yet for the contract in the period
```

Comment:
Imo it would be more readable in more than one method, and in general I think it is better a method that returns true or false with the message of what is doing in the name than a comment of why that comparison itself.
So something like this for example:
```ruby
def bill_plan_change(contract, period)
    plan = contract.plan
    old_plan = contract.old_plan

    bill_period(contract, period) unless fixed_cost_billed?(contract, period)

    add_plan_cost(:refund, contract, old_plan, period)
    add_plan_cost(:bill, contract, plan, period)
end

private

def bill_period(contract, period)
    old_plan_period_begin = [contract.paid_until.utc, period.begin.utc.beginning_of_month].max
    old_plan_period = TimeRange.new(old_plan_period_begin, period.end.utc)
    add_plan_cost(:bill, contract, old_plan, old_plan_period)
end

def fixed_cost_billed?(contract, period)
    contract.paid_until.utc >= period.begin.utc
end
```

Reply:
Prepaid has it https://github.com/3scale/porta/blob/587505f9971dbbf829e7671c9a488243c1be2b83/app/models/finance/prepaid_billing_strategy.rb#L121-L127

By the way isn't it the same notion of unbilled period and billed period for postpaid?

Reply:
It is not the same. `not_billed_yet?` actually means "never ever billed before?". That's what use for prepaid. For postpaid, we need "has it been billed for the period already?"